pkg/apis/openebs.io/v1alpha1/upgradetask_types.go

```diff
@@ -0,0 +1,225 @@
+	Options *ResourceOptions `json:"options,omitempty"`
+}
+
+// CStorPoolInstance is the ResourceType for cstor pool cluster
+type CStorPoolInstance struct {
+	// CSPCName contains the name of the storage pool claim to be upgraded
+	CSPIName string `json:"cspiName,omitempty"`
+	// Options can be used to change the default behaviour of upgrade
+	Options *ResourceOptions `json:"options,omitempty"`
+}
+
+// ResourceOptions provides additional options for a particular resource
+type ResourceOptions struct {
+	// IgnoreStepsOnError allows to ignore steps which failed
+	IgnoreStepsOnError []string `json:"ignoreStepsOnError,omitempty"`
+}
+
+// UpgradeTaskStatus provides status of a upgradeTask
+type UpgradeTaskStatus struct {
+	// Phase indicates if a upgradeTask is started, success or errored
+	Phase UpgradePhase `json:"phase,omitempty"`
+	// StartTime of Upgrade
+	StartTime metav1.Time `json:"startTime,omitempty"`
+	// CompletedTime of Upgrade
+	CompletedTime metav1.Time `json:"completedTime,omitempty"`
+	// UpgradeDetailedStatuses contains the list of statuses of each step
+	UpgradeDetailedStatuses []UpgradeDetailedStatuses `json:"upgradeDetailedStatuses,omitempty"`
+	// Retries is the number of times the job attempted to upgrade the resource
+	Retries int `json:"retries"`
+}
+
+// UpgradeDetailedStatuses represents the latest available observations
+// of a UpgradeTask current state.
+type UpgradeDetailedStatuses struct {
+	Step UpgradeStep `json:"step,omitempty"`
+	// StartTime of a UpgradeStep
+	StartTime metav1.Time `json:"startTime,omitempty"`
+	// LastUpdatedTime of a UpgradeStep
+	LastUpdatedTime metav1.Time `json:"lastUpdatedAt,omitempty"`
+	// Status of a UpgradeStep
+	Status `json:",inline"`
+}
+
+// Status represents the state of the step performed during the upgrade.
+type Status struct {
+	// Phase indicates if the UpgradeStep is waiting, errored or completed.
+	Phase StepPhase `json:"phase,omitempty"`
+	// A human-readable message indicating details about why the upgradeStep
+	// is in this state
+	Message string `json:"message,omitempty"`
+	// Reason is a brief CamelCase string that describes any failure and is meant
+	// for machine parsing and tidy display in the CLI
+	Reason string `json:"reason,omitempty"`
+}
+
+// UpgradeStep is the current step being performed for a particular resource upgrade
+type UpgradeStep string
+
+const (
+	// PreUpgrade is the step to verify resource before upgrade
+	PreUpgrade UpgradeStep = "PRE_UPGRADE"
+	// TargetUpgrade is the step to upgrade Target depoyment of resource
+	TargetUpgrade UpgradeStep = "TARGET_UPGRADE"
+	// ReplicaUpgrade is the step to upgrade replica deployment of resource
+	ReplicaUpgrade UpgradeStep = "REPLICA_UPGRADE"
+	// Verify is the step to verify the upgrade
+	Verify UpgradeStep = "VERIFY"
+	// Rollback is the step to rollback to previous version if upgrade fails
+	Rollback UpgradeStep = "ROLLBACK"
+	// PoolInstanceUpgrade is the step to verify resource before upgrade
```

Comment:
Can this comment be more detailed or explanatory, I didn't get a hold of it.

Reply:
Also, Is this is going to be the only new step that will be included for CSPC and CSPI upgrade status?

Reply:
This step is present in old upgrade also. As the pool does not have any target or replica terms attached to it we have this step for each CSP. The same can be used to represent CSPI upgrade.
https://github.com/openebs/maya/blob/e69d3976173625c05b09da74fcd86e491a0bee2a/pkg/upgrade/upgrader/csp_upgrade.go#L307